src/shadowbox/server/server_access_key.spec.ts

```diff
@@ -76,77 +84,279 @@ describe('ServerAccessKeyRepository', () => {
+        new ManagerMetricsStub({}));
+    // Check that repo1 and repo2 have the same access keys
+    const repo1Keys = repo1.listAccessKeys();
+    const repo2Keys = repo2.listAccessKeys();
+    expect(repo1Keys.length).toEqual(2);
+    expect(repo2Keys.length).toEqual(2);
+    expect(repo1Keys[0]).toEqual(repo2Keys[0]);
+    expect(repo1Keys[1]).toEqual(repo2Keys[1]);
+    done();
   });
 
   it('Does not re-use ids when using the same config file', (done) => {
     const config = new InMemoryConfig<AccessKeyConfigJson>({accessKeys: [], nextId: 0});
     // Create a repo with 1 access key, then delete that access key.
     const repo1 = new ServerAccessKeyRepository(
-        new PortProvider(), 'hostname', config, createMockShadowsocksServer());
+        new PortProvider(), '', config, new MockShadowsocksServer(), new ManagerMetricsStub({}));
     repo1.createNewAccessKey().then((accessKey1) => {
       repo1.removeAccessKey(accessKey1.id);
 
       // Create a 2nd repo with one access key, and verify that
       // it hasn't reused the first access key's ID.
       const repo2 = new ServerAccessKeyRepository(
-          new PortProvider(), 'hostname', config, createMockShadowsocksServer());
+          new PortProvider(), '', config, new MockShadowsocksServer(), new ManagerMetricsStub({}));
       repo2.createNewAccessKey().then((accessKey2) => {
         expect(accessKey1.id).not.toEqual(accessKey2.id);
         done();
       });
     });
   });
-});
 
-// Convert from an IterableIterator to an Array
-function iterToArray<T>(iter: IterableIterator<T>): T[] {
-  const returnArray = [];
-  for (const el of iter) {
-    returnArray.push(el);
-  }
-  return returnArray;
-}
+  it('start exposes the access keys to the server', async (done) => {
+    const config = new InMemoryConfig<AccessKeyConfigJson>({accessKeys: [], nextId: 0});
+    const repo = new ServerAccessKeyRepository(
+        new PortProvider(), '', config, new MockShadowsocksServer(), new ManagerMetricsStub({}));
+    const accessKey1 = await repo.createNewAccessKey();
+    const accessKey2 = await repo.createNewAccessKey();
+    // Create a new repository with the same configuration. The keys should not be exposed to the
+    // server until `start` is called.
+    const server = new MockShadowsocksServer();
+    const repo2 = new ServerAccessKeyRepository(
+        new PortProvider(), '', config, server, new ManagerMetricsStub({}));
+    expect(server.getAccessKeys().length).toEqual(0);
+    await repo2.start(new ManualClock());
+    const serverAccessKeys = server.getAccessKeys();
+    expect(serverAccessKeys.length).toEqual(2);
+    expect(serverAccessKeys[0].id).toEqual(accessKey1.id);
+    expect(serverAccessKeys[1].id).toEqual(accessKey2.id);
+    done();
+  });
 
-function countAccessKeys(repo: AccessKeyRepository) {
-  return iterToArray(repo.listAccessKeys()).length;
-}
+  it('start periodically enforces access key quotas', async (done) => {
```

Comment:
Please don't kill the `Clock` interface. Using the native interface makes it hard to work test. It's too magical, and magical is bad. It's a lot easier to understand the Clock interface.

We are "fast forwarding" we do that with the clock.runCallbacks(). Maybe we need to add a way to set the "now" if you wanna test in more details.

Reply:
Discussed offline, vini and I discussed an implementation like [this](http://www.typescriptlang.org/play/#code/MYGwhgzhAECyYDsCuYQGEQHtgGtoG8AoaE6ABwCcBLANzABcBTaYVEAIzFxgF5oBtALrRIBViBAAuaADEkCYPSqYEAGmhKAto2nJN7RhXVUETCnSnQ9Bo9CoATXUn2GAvkIDcxUpVoNmWjpWzjbQfAAMXqTk1HRMVowAHvQAkvZh0JGE3iQImADuABQAlE4uFATQFIz0SBQIGgAWVBAAdIEe0K7Z0b5xzBCYFPRobJzcJQQ50U0treIcXDi8s20L48vTM6Stg8OFhcAAjOrAAEzFYQB8LEftVNrQALQsZ-faxVGk3dMQNSmmQwWQqBWAAZTKNnUADNpHIFEoVKVguUptsSPRmmsxks2mQkBBGoV8AtpND1IFpJi5oFoABqDQPRjg4yA8yoKlMll2RyrVoIJKpex0umuT5bUjVWr1PkC5JpZ7QI5fEg-aJ-egAFSZmCQ9BBXIhKKh0FhsnkimUCGR1kMaPR1OxEg2eIJRJJbDJFKZVKx72YDNBYNZZgs0gB0JMVHoAE9jLzHfzBWkRWKVTMpXUGom5ULFcrpmrSEpcAbtODIYZLtIaJgHPbtlRodAy8ywdAADwRS6YigFaAAJUQAHNGABRCh9iiFABE2twjMemgJ9GgBnImAg0dojBn4vRdmbIL961xrRAjAQw8xYW7DYPfNpdL4QfTB8z9Tf3x66OAKggq6aGAiTao8L5+k+i5tl4EoYn6ewjDiEz7uiF6rrmGSJqe3D8OEggwQe+TNBeLa5v6nZ8EBIFMpcRAPsWEFMhkZEdLBMxkQsJRftsLFMc+CTJK0Jihqg3EzImCGjM6uJcWx0QYeBczYcsuH4WxRbbImtKUcBoGMOm3Q-GhLDMYw+RwIgKDoFgpbisAuz-GywIAKzqJMPA3H+CCDBe56YMOhTOcUdkOakTmoIUACcbmXB5LD-pgvlYAFkXBV49kanpur6mcADMMXXPF3mJYwfkBXlaWEBlNRZXqgUnC2sWeQlSX+fVwVVfcpbOeEIWZTqdWuY1hVeT5pXJYFQUhSWOCFLl01UKW80eEAA) to allow fast forwarding

Reply:
Jon's manual Clock is nice, and it would be good to add to our codebase and use it in the test. But I'd be fine doing that in another PR, to get this big one out of the way and unblock the other changes we want to do.

Reply:
Cool - I'd rather do it in a follow-up PR.